stats: support multiple repos

stefanhengl opened the pull request:
This updates calculateStats to support multiple repos. The idea is to
compute the file range for each repo and then compute the stats per file
range. A special case is `memoryUse()` which will still return data for
the entire shard.

stefanhengl committed:
stats: support multiple repos

This updates calculateStats to support multiple repos. The idea is to
compute the file range for each repo and then compute the stats per file
range. A special case is `memoryUse()` which will still return data for
the entire shard.

stefanhengl committed:
fix lastFN

keegancsmith commented on build/e2e_test.go:
```diff
@@ -464,6 +464,9 @@ func TestEmptyContent(t *testing.T) {
 	}
+	b.Add(zoekt.Document{
+		Name: "main.go",
+	})
```
does this mean we don't support empty repos now?

keegancsmith replied:
I updated so we support it.

keegancsmith commented on indexdata.go:
```diff
@@ -157,25 +157,29 @@ func (d *indexData) getChecksum(idx uint32) []byte {
-func (d *indexData) calculateStatsForRepoIndex(i int) RepoListEntry {
+// calculates stats for files in the range [start, end). It is assumed that all
+// files in the range belong to the same repository.
+func (d *indexData) calculateStatsForFileRange(start, end uint32) RepoListEntry {
```
for the merge process we need to check the invariant that repos are over contiguous ranges. Should we just represent it that way, instead of keeping the invariant?

keegancsmith commented on indexdata.go:
```diff
@@ -184,26 +188,32 @@ func (d *indexData) calculateStatsForRepoIndex(i int) RepoListEntry {
+			end++
+		}
+		d.repoListEntry = append(d.repoListEntry, d.calculateStatsForFileRange(start, end))
+		start = end
```
your baking in the invariant htat data will appear in the same order as repoListEntries. I don't think we have this assumption anywhere else?

keegancsmith replied:
I added an explicit cxheck for this and returned an error if we don't

keegancsmith commented on indexdata.go:
```diff
@@ -184,26 +188,32 @@ func (d *indexData) calculateStatsForRepoIndex(i int) RepoListEntry {
+// time. However, we find these values very useful to track and computing them
+// outside of load time introduces a lot of complexity.
+func (d *indexData) calculateNewLinesStats(start, end uint32) (count, defaultCount, otherCount uint64) {
+	for i, branchMask := range d.fileBranchMasks[start:end] {
```
bug: i is no longer correct anymore. it is used below to access newlinesIndex.

keegancsmith reviewed (changes requested):
Cool stuff. One blocking change. I'll maybe pick this up depending on when ya get back.

keegancsmith commented:
Must remember to stop ignoring documents as was added to `TestShardedSearcher_List`: https://github.com/sourcegraph/zoekt/pull/99/files#diff-76a58fecae84e38d1a419bf30150d63ed91fadc5b731723a06b3cb0fa6224f0dR385

keegancsmith committed:
indexdata: support empty shards for listing

Adjusted calculateStatsForFileRange to only return RepoStats. This way
we can still add entries for repos with no documents. Also added some
inline code review comments for Stefan.

Change-Id: I2d043bd6bad8b216cd2f61998c4b242957307a4c

keegancsmith committed:
indexdata: use correct docID in calculateNewLinesStats

Using a slice of fileBranchMasks lead to "i" starting at "0" instead of
"start".

Change-Id: I7a83f5200e7fa07cb86eafca8d781275bf80ed7a

keegancsmith committed:
indexdata: validate invariant that documents are in order of repos

Change-Id: Ibda2aaead7388b86eb02a7283579f4a95acd9cce
